src/qt/bantablemodel.cpp

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2011-2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "bantablemodel.h"
+
+#include "clientmodel.h"
+#include "guiconstants.h"
+#include "guiutil.h"
+
+#include "net.h"
+#include "sync.h"
+#include "utiltime.h"
+
+#include <QDebug>
+#include <QList>
+#include <QTimer>
+
+#include <boost/date_time/posix_time/posix_time.hpp>
+#include <boost/date_time/c_local_time_adjustor.hpp>
+
+// private implementation
+class BanTablePriv
+{
+public:
+    /** Local cache of peer information */
+    QList<CCombinedBan> cachedBanlist;
+    /** Column to sort nodes by */
+    int sortColumn;
+    /** Order (ascending or descending) to sort nodes by */
+    Qt::SortOrder sortOrder;
+
+    /** Pull a full list of banned nodes from CNode into our cache */
+    void refreshBanlist()
+    {
+        std::map<CSubNet, int64_t> banMap;
+        CNode::GetBanned(banMap);
+
+        cachedBanlist.clear();
+#if QT_VERSION >= 0x040700
+        cachedBanlist.reserve(banMap.size());
+#endif
+        std::map<CSubNet, int64_t>::iterator iter;
+        for (iter = banMap.begin(); iter != banMap.end(); ++iter) {
+            CCombinedBan banEntry;
+            banEntry.subnet = iter->first;
+            banEntry.bantil = iter->second;
+            cachedBanlist.append(banEntry);
+        }
+    }
+
+    int size()
+    {
+        return cachedBanlist.size();
+    }
+
+    CCombinedBan *index(int idx)
+    {
+        if(idx >= 0 && idx < cachedBanlist.size()) {
+            return &cachedBanlist[idx];
+        } else {
```

Comment:
Not sure if this else can be removed. What if the `QTableView` requests a row at index which is not in the cache `QList`. Obviously it should not happen because the table size is driven by the size of the cache `QList`. But I prefer keeping this.